### PR TITLE
fix: issue with missing start/end time and handling excepts for end t…

### PR DESCRIPTION
**Pull request by willbakst**
- Added better handling for setting start and end time for extract and errors
- Added back start and end time for completions (so you can see diff between construction, calling, return, etc.)

### CODE REVIEW

**Review by brianpark0126 — APPROVED** [review]
lgtm

File: mirascope/openai/prompt.py [hunk 1]
```diff
@@ -308,22 +337,33 @@ async def async_extract(self, schema, retries=0):
             openai_tool = OpenAITool.from_model(schema)
             return_tool = False
 
-        completion = await self.async_create(tools=[openai_tool])
+        try:
```
Comment (brenkao) [inline mirascope/openai/prompt.py:340]: define self._start_time here? then you can remove 342 and 344 and 328

File: mirascope/openai/prompt.py [hunk 2]
```diff
@@ -308,22 +337,33 @@ async def async_extract(self, schema, retries=0):
...
+            completion = await self.async_create(tools=[openai_tool])
+            self._start_time = extraction_start_time
+        except Exception:
+            self._end_time = datetime.datetime.now().timestamp() * 1000
+            self._start_time = extraction_start_time
+            raise
         try:
             tool = completion.tool
             if tool is None:
+                self._end_time = datetime.datetime.now().timestamp() * 1000
```
Comment (brenkao) [inline mirascope/openai/prompt.py:350]: why do all these end times need to be defined in each conditional. shouldnt end time be when completion is complete?

File: mirascope/openai/prompt.py (hunk 1 above)
Comment (willbakst) [reply mirascope/openai/prompt.py:340]: create and async_create also set start time, but that one doesn't take into account all of the time spend converting tools, so I did it this way

File: mirascope/openai/prompt.py (hunk 2 above)
Comment (willbakst) [reply mirascope/openai/prompt.py:350]: These are all above return statements, so it's right when it completes. The completion also has _start_time and _end_time for when it completes

**Review by brianpark0126 — COMMENTED** [review]
encountered some errors when I just ran the code, essentially completion.dump() always gives None for start_time and end_time, and then theres an issue where when you do prompt.dump(completion.dump()) there's a doubling up of start_time and start_time_ms and same for end times, except start_time and end_time, are None. Snippet with outputs structured for reference: 
```
import os

# import openai api key from .env
from dotenv import load_dotenv

from mirascope import tags
from mirascope.openai import OpenAIPrompt

load_dotenv()


@tags(["recommendation_project", "version:0001"])
class BookRecommendationPrompt(OpenAIPrompt):
    """
    Can you recommend some books on {topic}?
    """

    topic: str


prompt = BookRecommendationPrompt(topic="how to bake a cake")
comp = prompt.create()
print(comp.dump(), "\n\n")
print(prompt.dump(comp.dump()))
output1 = {
    "start_time": None,
    "end_time": None,
    "output": {
        "id": "chatcmpl-90FfY1MlXcuiwYITvNcy9RjASDB47",
        "choices": [
            {
                "finish_reason": "stop",
                "index": 0,
                "logprobs": None,
                "message": {
                    "content": '1. "The Cake Bible" by Rose Levy Beranbaum - ...
                    "role": "assistant",
                    "function_call": None,
                    "tool_calls": None,
                },
            }
        ],
        "created": 1709847260,
        "model": "gpt-3.5-turbo-0125",
        "object": "chat.completion",
        "system_fingerprint": "fp_b9d4cef803",
        "usage": {"completion_tokens": 176, "prompt_tokens": 19, "total_tokens": 195},
    },
}

output2 = {
    "template": "Can you recommend some books on {topic}?",
    "inputs": {"api_key": None, "topic": "how to bake a cake"},
    "tags": ["recommendation_project", "version:0001"],
    "call_params": {"model": "gpt-3.5-turbo-0125"},
    "start_time_ms": 1709847488599.576,
    "end_time_ms": 1709847491881.78,
    "start_time": None,
    "end_time": None,
    "output": {
        "id": "chatcmpl-90FjE0iXj8CY7i4eDbHkgHONRVuJi",
        "choices": [
            {
                "finish_reason": "stop",
                "index": 0,
                "logprobs": None,
                "message": {
                    "content": '1. "The Cake Bible" by Rose Levy Beranbaum\n2. ...
                    "role": "assistant",
                    "function_call": None,
                    "tool_calls": None,
                },
            }
        ],
        "created": 1709847488,
        "model": "gpt-3.5-turbo-0125",
        "object": "chat.completion",
        "system_fingerprint": "fp_2b778c6b35",
        "usage": {"completion_tokens": 140, "prompt_tokens": 19, "total_tokens": 159},
    },
}
```

File: mirascope/openai/prompt.py (hunk 1 above)
Comment (brenkao) [reply mirascope/openai/prompt.py:340]: we are timing the actual completion, not the time it takes to do conversion.

File: mirascope/openai/prompt.py (hunk 2 above)
Comment (brenkao) [reply mirascope/openai/prompt.py:350]: i understand that, but the end time should be set when the completion finishes, not when you hit the conditional

Comment (willbakst) [reply mirascope/openai/prompt.py]: Resolved, only completion has it now.

File: mirascope/openai/prompt.py (hunk 1 above)
Comment (willbakst) [reply mirascope/openai/prompt.py]: Only completion has it now

**Comment by willbakst** [comment]
The mentioned bug should no longer be an issue. Start/end time available in the completion only as required parameters

**Review by brianpark0126 — APPROVED** [review]

**Pull request merged**